File: backend/bluffword-backend/src/main/java/org/bluffwordbackend/controllers/GameRoomController.java
package org.bluffwordbackend.controllers;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.bluffwordbackend.dtos.GameRoomSettingsDto;
import org.bluffwordbackend.dtos.JoinGameRoomRequestDto;
import org.bluffwordbackend.redisDtos.PlayerDto;
import org.bluffwordbackend.services.GameRoomBroadcaster;
import org.bluffwordbackend.services.GameRoomService;
import org.bluffwordbackend.services.RoundService;
import org.bluffwordbackend.websocket.WebSocketDisconnectListener;
import org.modelmapper.ModelMapper;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.messaging.handler.annotation.DestinationVariable;
import org.springframework.messaging.handler.annotation.MessageMapping;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@Slf4j
@RestController
@RequiredArgsConstructor
@RequestMapping("api/gameRoom")
@CrossOrigin(origins = "http://localhost:5173")
public class GameRoomController {

   private final SimpMessagingTemplate messagingTemplate;
   private final WebSocketDisconnectListener webSocketDisconnectListener;
    private final GameRoomService gameRoomService;
    private final RoundService roundService;

    private final GameRoomBroadcaster gameRoomBroadcaster;

    private final ModelMapper modelMapper;


    @GetMapping("/test")
    public ResponseEntity<?> getListOfPlayers(@RequestParam String roomCode) {
        List<PlayerDto> playerDtos = gameRoomService.getListOfPlayers(roomCode);
        return ResponseEntity.ok(playerDtos);
    }

    @PostMapping("/create")
    public ResponseEntity<?> createRoom(@RequestBody String nickname, @RequestBody GameRoomSettingsDto gameRoomSettingsDto) {
        if ( nickname == null || nickname.isBlank()) {
                        return ResponseEntity.badRequest().body(Map.of("message", "nickname is required"));
        }
        if ( gameRoomSettingsDto == null ) {
            return ResponseEntity.badRequest().body(Map.of("message", "game Settings are required"));
        }

        PlayerDto playerDto = gameRoomService.createRoom(nickname, gameRoomSettingsDto);
        if (playerDto == null) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }

        List<PlayerDto> playerDtos = gameRoomService.getListOfPlayers(playerDto.getRoomCode());
        gameRoomBroadcaster.broadcastPlayers(playerDto.getRoomCode(), playerDtos);

        return ResponseEntity.ok(playerDto);
    }

    @PostMapping("/join")
    public ResponseEntity<?> join(@RequestBody JoinGameRoomRequestDto request) {
        PlayerDto playerDto = gameRoomService.joinRoom(request.nickname(), request.roomCode());
        List<PlayerDto> playerDtos = gameRoomService.getListOfPlayers(playerDto.getRoomCode());
        gameRoomBroadcaster.broadcastPlayers(playerDto.getRoomCode(), playerDtos);

        return ResponseEntity.ok(playerDto);
    }


    @MessageMapping("/room/{roomCode}/players")
    public void getPlayers(@DestinationVariable String roomCode) {
        List<PlayerDto> playerDtos = gameRoomService.getListOfPlayers(roomCode);
        gameRoomBroadcaster.broadcastPlayers(roomCode, playerDtos);

    }

    @DeleteMapping("/players")
    public void deletePlayerFromRoom(@RequestBody Map<String, String> request) {
        String code = request.get("roomCode");
        Long playerId = Long.parseLong(request.get("playerId"));
        gameRoomService.deletePlayerFromRoom(code, playerId);

        List<PlayerDto> playerDtos = gameRoomService.getListOfPlayers(code);
        gameRoomBroadcaster.broadcastPlayers(code, playerDtos);
    }

    @GetMapping("/settings")
    public ResponseEntity<?> getGameSettings(@RequestParam String roomCode) {

       return ResponseEntity.ok(gameRoomService.getRoomSettings(roomCode));

    }


}

File: backend_kotlin/src/main/kotlin/org/project/backend_kotlin/gameRoom/GameRoomController.kt
package org.project.backend_kotlin.gameRoom

import org.project.backend_kotlin.gameRoom.dto.CreateRoomRequestDto
import org.project.backend_kotlin.gameRoom.dto.JoinRoomRequestDto
import org.project.backend_kotlin.redisModels.Player
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import jakarta.validation.Valid
import jakarta.validation.constraints.NotEmpty
import org.project.backend_kotlin.redisModels.GameRoom
import org.springframework.messaging.handler.annotation.DestinationVariable
import org.springframework.messaging.handler.annotation.MessageMapping
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestParam

@RestController
@RequestMapping("/api/gameRoom")
class GameRoomController (
    private val gameRoomManagementService: GameRoomManagementService,
    private val gameRoomBroadcaster: GameRoomBroadcaster,

    ) {


    @PostMapping("/create")
    fun createGameRoom(@RequestBody @Valid createRoomDto: CreateRoomRequestDto) : ResponseEntity<Player> {

        val createdPlayer = gameRoomManagementService.createGameRoom(createRoomDto.nickname, createRoomDto.settings)
        println(createdPlayer)
        return ResponseEntity.ok(createdPlayer)

    }

    @PostMapping("/join")
    fun joinGameRoom(@RequestBody @Valid joinRoomRequestDto: JoinRoomRequestDto) : ResponseEntity<Player> {

        return ResponseEntity.ok(gameRoomManagementService.joinGameRoom(joinRoomRequestDto.roomCode, joinRoomRequestDto.nickname))
    }

    @MessageMapping("/room/{roomCode}/players")
    fun getPlayersFromRoom(@DestinationVariable roomCode: String) {
        println(gameRoomManagementService.getPlayers(roomCode))

        gameRoomBroadcaster.broadcastPlayers(roomCode, gameRoomManagementService.getPlayers(roomCode))
    }

    @MessageMapping("/room/{roomCode}/state")
    fun getRoomState(@NotEmpty @DestinationVariable roomCode: String) =
        gameRoomBroadcaster.broadcastGameRoomState(roomCode, gameRoomManagementService.getRoomState(roomCode))

    @DeleteMapping("/player")
    fun deletePlayerFromRoom(@RequestBody req: JoinRoomRequestDto) {
        // to change from nickname to playerId
        gameRoomManagementService.deletePlayerFromRoom(req.roomCode, req.nickname)
        gameRoomBroadcaster.broadcastPlayers(req.roomCode, gameRoomManagementService.getPlayers(req.roomCode))
    }

    @GetMapping("/settings")
    fun getGameSettings(@RequestParam roomCode: String): ResponseEntity<GameRoom> =
         ResponseEntity.ok(gameRoomManagementService.getGameRoomConfig(roomCode))

}